### ISSUE TITLE
add features to `make_afun` 

### ISSUE THREAD

**Issue by danielinteractive** [Roche/rtables#77]
Thanks @gmbecker for `make_afun`, it looks pretty good already. However, in order to switch to using it in production, we would still need the following features for `make_afun`:

- I think this is more a bug, but also listing here: The labels that are passed from the statistics function result are not ending up in the table as they should. Example:
	```r
	s_summary <- function(x) {
	  stopifnot(is.numeric(x))
	
	  list(
	    n = with_label(sum(!is.na(x)), "N subjects"),
	    mean_sd = with_label(c(mean = mean(x), sd = sd(x)), "My mean and SD"),
	    min_max = with_label(range(x), "Range")
	  )
	}
	a_summary <- make_afun(
	  s_summary,
	  .labels = c(n = "n subjects"),  # only overwrite the label of the `n` statistics here.
	  .formats = c(n = "xx", mean_sd = "xx.xx (xx.xx)", min_max = "xx.xx - xx.xx")
	)
	basic_table() %>%
	  analyze(
	    "Sepal.Length",
	    afun = a_summary
	  ) %>%
	  build_table(iris)
	```
	Here we have labels "n", "mean_sd" and "min_max" instead of the expected "n subjects", "My mean and SD" and "Range".

- We need `indent_mods` functionality for `make_afun`, analogous to `formats` functionality (only nonnegative indent modification would be sufficient, compare current prototype we have internally in `tern`)

- I think it is nice to be able to change default arguments of the statistics function. However, would it be possible to make the logical step to also allow to set a new default for arguments that did not have a default earlier in the statistics function? Example:
	```r
	s_foo <- function(df, .N_col, a = 1, b = 2) {
	   list(
	      nrow_df = nrow(df),
	      .N_col = .N_col,
	      a = a,
	      b = b
	   )
	}
	a_foo <- make_afun(s_foo, .N_col = 40,
	 .formats = c(nrow_df = "xx.xx", ".N_col" = "xx.", a = "xx", b = "xx.x"),
	 .labels = c(nrow_df = "Nrow df", ".N_col" = "n in cols", a = "a value", b = "b value")
	)
	a_foo(iris)
	```
	does not work currently.

- When handling reference columns we need to be able to pass empty strings instead of
numbers from a statistics function. In that case, the `afun` returned by `make_afun` should handle that case by changing the format to `"xx"` and not use the usually numeric format which will fail in the table production step. (I could imagine that this might be also solved by modifying the general interpretation of formats to not do anything when a string is detected.) Note that we cannot work with `NULL` instead of strings because that does not allow attributes which we need for passing labels.

	Example:
	```r
	s_summary_ref <- function(x, .in_ref_col) {
	  list(
        n = with_label(`if`(in_ref_col, "", length(x)), "Number of patients"),
        mean = with_label(`if`(in_ref_col, "", mean(x)), "Mean"),
        median = with_label(`if`(in_ref_col, "", median(x)), "Median")
      )
	}
	a_summary_ref <- make_afun(
	  s_summary_ref,
	  .formats = c(n = "xx", mean_sd = "xx.xx (xx.xx)", min_max = "xx.xx - xx.xx")
	)
	basic_table() %>%
	  split_cols_by("Species", ref_group = "setosa") %>%
	  analyze("Sepal.Length", afun = a_summary_ref) %>%
	  build_table(iris)
	```

**Comment by danielinteractive** [comment]
@waddella fyi

**Comment by gmbecker** [comment]
Hi @danielinteractive 

I will address the other points separately later, but re overriding default values:

> * I think it is nice to be able to change default arguments of the statistics function.

So it doesn't actually change the defaults, per se, in the traditional sense (we could change this, I suppose but it woulud require even more language hackery so i'd prefer not to). It is more accurate to say that it uses a separate mechanism internally to override the default value for arguments you passed to the constructor. (This is a distinction largely without a difference in practice, but I wanted to be explicit incase it did come up someday).

> However, would it be possible to make the logical step to also allow to set a new default for arguments that did not have a default earlier in the statistics function? Example:
>   ```r
>    s_foo <- function(df, .N_col, a = 1, b = 2) {
>       list(
>          nrow_df = nrow(df),
>          .N_col = .N_col,
>          a = a,
>          b = b
>       )
>    }
>    a_foo <- make_afun(s_foo, .N_col = 40,
>     .formats = c(nrow_df = "xx.xx", ".N_col" = "xx.", a = "xx", b = "xx.x"),
>     .labels = c(nrow_df = "Nrow df", ".N_col" = "n in cols", a = "a value", b = "b value")
>    )
>    a_foo(iris)
>   ```
>   
>   
>   does not work currently.

So what is happening her eis actually something else. You cannot override (or "set") the default value for `.N_col` or any of the "framework internal" type arguments (`.N_total`, `.ref_group`, `.in_ref_col`) because those will be passed to the function explicitly by the tabulation machinery and this will not override that. 

If that works now when `.N_col` has a default value that is a bug and it *will not work in the future*. The way to customize the`.N_col` that your analysis funs receive is to override the column Ns for the table.

Note this currently works for arguments that are not used by the internal tabulation machinery explicitly:

```
s_nodflt <- function(df, .N_col, a = 1, b) {
        list(
            nrow_df = nrow(df),
       .N_col = .N_col,
       a = a,
       b = b
       )
    }

    a_nodflt <- make_afun(s_nodflt, b = 4,
                          .formats = c(nrow_df = "xx.xx", ".N_col" = "xx.", a = "xx", b = "xx.x"),
                          .labels = c(nrow_df = "Nrow df", ".N_col" = "n in cols", a = "a value", b = "b value")
                          )
 a_nodflt(iris, 5)[["b"]]
```

gives
```
[[1]]
[1] 4

attr(,"format")
[1] "xx.x"
attr(,"colspan")
[1] 1
attr(,"label")
[1] "b value"
attr(,"class")
[1] "CellValue"
```

**Comment by danielinteractive** [comment]
Thanks @gmbecker ! Cool, given that this works for all non-rtables-specific arguments it is fine as is. So no need to change anything here around the arguments default-ing behavior from my side.

**Comment by gmbecker** [comment]
> Thanks @gmbecker for `make_afun`, it looks pretty good already. However, in order to switch to using it in production, we would still need the following features for `make_afun`:
> 
> * I think this is more a bug, but also listing here: The labels that are passed from the statistics function result are not ending up in the table as they should. 

Fixed.

> * We need `indent_mods` functionality for `make_afun`, analogous to `formats` functionality (only nonnegative indent modification would be sufficient, compare current prototype we have internally in `tern`)

Implemented in 027e0e9

> * I think it is nice to be able to change default arguments of the statistics function. However, would it be possible to make the logical step to also allow to set a new default for arguments that did not have a default earlier in the statistics function? 

Always worked for non-rtables-internal arguments.

> * When handling reference columns we need to be able to pass empty strings instead of
>   numbers from a statistics function. In that case, the `afun` returned by `make_afun` should handle that case by changing the format to `"xx"` and not use the usually numeric format which will fail in the table production step. (I could imagine that this might be also solved by modifying the general interpretation of formats to not do anything when a string is detected.) Note that we cannot work with `NULL` instead of strings because that does not allow attributes which we need for passing labels.

There are two options here. One, please see `non_ref_rcell` in ?rcell and the corresponding `.null_ref_cells` argument to `make_afun` (which internally makes use of it). Via this method you would return whatever non-NULL value you want, but it will be nulled by the machinery itself. 

The second is that returning character() instead of "" works now as intended, as far as I know.

A third point is that it still seems *really* wrong to have it factored this way where labels are a concept outside of cells. Its a shortcut to heartbreak and tragedy. We should talk about the reason this is such a hard requirement more.

That said, I believe this closes the issue (at least for now) @danielinteractive @waddella

**Comment by danielinteractive** [comment]
Thanks a lot @gmbecker. This looks all pretty good, I just tested it and only so far came across the one small default argument bug which I filed separately. So this is closed in terms of features.
